types/canonical.go

```diff
@@ -0,0 +1,116 @@
+package types
+
+import (
+	"time"
+
+	cmn "github.com/tendermint/tendermint/libs/common"
+	tmtime "github.com/tendermint/tendermint/types/time"
+)
+
+// Canonical* wraps the structs in types for amino encoding them for use in SignBytes / the Signable interface.
+
+// TimeFormat is used for generating the sigs
+const TimeFormat = time.RFC3339Nano
+
+type CanonicalBlockID struct {
+	Hash        cmn.HexBytes           `json:"hash,omitempty"`
+	PartsHeader CanonicalPartSetHeader `json:"parts,omitempty"`
+}
+
+type CanonicalPartSetHeader struct {
+	Hash  cmn.HexBytes `json:"hash,omitempty"`
+	Total int          `json:"total,omitempty"`
+}
+
+type CanonicalProposal struct {
+	ChainID          string                 `json:"@chain_id"`
+	Type             string                 `json:"@type"`
+	BlockPartsHeader CanonicalPartSetHeader `json:"block_parts_header"`
+	Height           int64                  `json:"height"`
+	POLBlockID       CanonicalBlockID       `json:"pol_block_id"`
+	POLRound         int                    `json:"pol_round"`
+	Round            int                    `json:"round"`
+	Timestamp        time.Time              `json:"timestamp"`
+}
+
+type CanonicalVote struct {
+	ChainID   string           `json:"@chain_id"`
+	Type      string           `json:"@type"`
+	BlockID   CanonicalBlockID `json:"block_id"`
+	Height    int64            `json:"height"`
+	Round     int              `json:"round"`
+	Timestamp time.Time        `json:"timestamp"`
+	VoteType  byte             `json:"type"`
+}
```

Comment:
I understand that this might be a first step towards solving https://github.com/tendermint/tendermint/issues/1622

This is very different from the agreed schema  (https://github.com/tendermint/tendermint/issues/1622#issuecomment-422363861)
```
// vanilla protobuf / amino encoded
message Vote {
    Version       fixed32                      
    Height        sfixed64       
    Round         sfixed32
    VoteType      fixed32
    Timestamp     Timestamp         // << using protobuf definition
    BlockID       BlockID           // << as already defined 
    ChainID       string            // at the end because length could vary a lot
}
```
The order and fixed size types are important so some fields can be serialized without having to deal with amino. This has been extensively discussed in https://github.com/tendermint/tendermint/issues/1622